ecc/bls12-377/pairing.go

```diff
@@ -341,3 +342,231 @@ func (p *g2Proj) lineCompute(evaluations *lineEvaluation, a *G2Affine) {
+// ----------------------
+
+type LineEvaluationAff struct {
+	R0 fptower.E2
+	R1 fptower.E2
+}
+
+// PairFixedQ calculates the reduced pairing for a set of points
+// ∏ᵢ e(Pᵢ, Qᵢ) where Q are fixed points in G2.
+//
+// This function doesn't check that the inputs are in the correct subgroup. See IsInSubGroup.
+func PairFixedQ(P []G1Affine, lines [][2][len(LoopCounter) - 1]LineEvaluationAff) (GT, error) {
+	f, err := MillerLoopFixedQ(P, lines)
+	if err != nil {
+		return GT{}, err
+	}
+	return FinalExponentiation(&f), nil
+}
+
+// PairingCheckFixedQ calculates the reduced pairing for a set of points and returns True if the result is One
+// ∏ᵢ e(Pᵢ, Qᵢ) =? 1 where Q are fixed points in G2.
+//
+// This function doesn't check that the inputs are in the correct subgroup. See IsInSubGroup.
+func PairingCheckFixedQ(P []G1Affine, lines [][2][len(LoopCounter) - 1]LineEvaluationAff) (bool, error) {
+	f, err := PairFixedQ(P, lines)
+	if err != nil {
+		return false, err
+	}
+	var one GT
+	one.SetOne()
+	return f.Equal(&one), nil
+}
+
+// PrecomputeLines precomputes the lines for the fixed-argument Miller loop
+func PrecomputeLines(Q G2Affine) (PrecomputedLines [2][len(LoopCounter) - 1]LineEvaluationAff) {
+	var accQ G2Affine
+	accQ.Set(&Q)
+
+	for i := len(LoopCounter) - 2; i >= 0; i-- {
+		accQ.doubleStep(&PrecomputedLines[0][i])
+		if LoopCounter[i] == 0 {
+			continue
+		} else {
+			accQ.addStep(&PrecomputedLines[1][i], &Q)
+		}
+	}
+	return PrecomputedLines
+}
+
+// MillerLoopFixedQ computes the multi-Miller loop as in MillerLoop
+// but Qᵢ are fixed points in G2 known in advance.
+func MillerLoopFixedQ(P []G1Affine, lines [][2][len(LoopCounter) - 1]LineEvaluationAff) (GT, error) {
+	n := len(P)
+	if n == 0 || n != len(lines) {
+		return GT{}, errors.New("invalid inputs sizes")
+	}
+
+	// no need to filter infinity points:
+	// 		1. if Pᵢ=(0,0) then -x/y=1/y=0 by gnark-crypto convention and so
+	// 		lines R0 and R1 are 0. At the end it happens that result will stay
+	// 		1 through the Miller loop because MulBy34(1,0,0)==1
+	// 		Mul34By34(1,0,0,1,0,0)==1 and MulBy01234(1,0,0,0,0)==1.
+	//
+	// 		2. if Qᵢ=(0,0) then PrecomputeLines(Qᵢ) will return lines R0 and R1
+	// 		that are 0 because of gnark-convention (*/0==0) in doubleStep and
+	// 		addStep. Similarly to Pᵢ=(0,0) it happens that result stays 1
+	// 		throughout the MillerLoop.
+
+	// precomputations
+	yInv := make([]fp.Element, n)
+	xNegOverY := make([]fp.Element, n)
+	for k := 0; k < n; k++ {
+		yInv[k].Inverse(&P[k].Y)
+		xNegOverY[k].Mul(&P[k].X, &yInv[k]).
+			Neg(&xNegOverY[k])
+	}
+
+	var result GT
+	result.SetOne()
+	var prodLines [5]E2
+
+	// Compute ∏ᵢ { fᵢ_{x₀,Q}(P) }
+	if n >= 1 {
+		// i = 62, separately to avoid an E12 Square
+		// (Square(res) = 1² = 1)
+		// LoopCounter[62] = 0
+		// k = 0, separately to avoid MulBy34 (res × ℓ)
+		// (assign line to res)
+
+		// line evaluation at P[0] (assign)
+		result.C1.B0.MulByElement(&lines[0][0][62].R0, &xNegOverY[0])
```

Comment:
I don't understand this part (the -x/y), it seems that you evaluate the line aX+bY+C=0 (in affine) as  a(-x/y) + b + c/y, why don't you compute it as ax+by+c ?

Reply:
This way you can force b to be 1. It is actually a’(-x/y)+1+c’/y=0 so that when you can specialize the multiplication by lines in Fpk (mulby034/014) to take account that one coefficient is 1, only two coefficients are nonzero and the rest is 0.